Remotion/Data/DomainObjects/Mapping/TypeDefinition.cs

```diff
@@ -0,0 +1,248 @@
+
+    public Type? StorageGroupType { get; }
+
+    public DefaultStorageClass DefaultStorageClass { get; }
+
+    protected TypeDefinition (Type type, Type? storageGroupType, DefaultStorageClass defaultStorageClass)
+    {
+      ArgumentUtility.CheckNotNull(nameof(type), type);
+
+      Type = type;
+      StorageGroupType = storageGroupType;
+      DefaultStorageClass = defaultStorageClass;
+
+      PropertyAccessorDataCache = new PropertyAccessorDataCache(this);
+      _cachedRelationEndPointDefinitions = new Lazy<RelationEndPointDefinitionCollection>(
+           () => RelationEndPointDefinitionCollection.CreateForAllRelationEndPoints(this, true),
+           LazyThreadSafetyMode.ExecutionAndPublication);
+      _cachedPropertyDefinitions =
+          new Lazy<PropertyDefinitionCollection>(
+              () => new PropertyDefinitionCollection(PropertyDefinitionCollection.CreateForAllProperties(this, true), true),
+              LazyThreadSafetyMode.ExecutionAndPublication);
+    }
+
+    public IStorageEntityDefinition StorageEntityDefinition
+    {
+      get
+      {
+        if (_storageEntityDefinition == null)
+          throw new InvalidOperationException($"StorageEntityDefinition has not been set for type '{Type.GetFullNameSafe()}'.");
+        return _storageEntityDefinition;
+      }
+    }
+
+    public bool HasStorageEntityDefinitionBeenSet => _storageEntityDefinition != null;
+
+    public PropertyDefinitionCollection MyPropertyDefinitions
+    {
+      get
+      {
+        if (_propertyDefinitions == null)
+          throw new InvalidOperationException($"No property definitions have been set for type '{Type.GetFullNameSafe()}'.");
+
+        return _propertyDefinitions;
+      }
+    }
+
+    public RelationEndPointDefinitionCollection MyRelationEndPointDefinitions
+    {
+      get
+      {
+        if (_relationEndPoints == null)
+          throw new InvalidOperationException($"No relation end point definitions have been set for type '{Type.GetFullNameSafe()}'.");
+
+        return _relationEndPoints;
+      }
+    }
+
+    public virtual bool IsTypeResolved => true;
+
+    public abstract bool IsPartOfInheritanceHierarchy { get; }
+
+    public abstract TypeDefinition GetInheritanceRoot ();
+
+    public PropertyDefinition? this [string propertyName]
+    {
+      get
+      {
+        ArgumentUtility.CheckNotNullOrEmpty(nameof(propertyName), propertyName);
+        return MyPropertyDefinitions[propertyName];
+      }
+    }
+
+    /// <summary>
+    /// Returns all types that are part of this type's hierarchy, including itself, using depth-first.
+    /// </summary>
+    public abstract TypeDefinition[] GetTypeHierarchy ();
+
+    protected abstract void CheckPropertyDefinitions (IEnumerable<PropertyDefinition> propertyDefinitions);
+
+    protected abstract void CheckRelationEndPointDefinitions (IEnumerable<IRelationEndPointDefinition> relationEndPoints);
```

Comment:
I personally like to retain the grouping of properties/methods and have the abstract members at the top of their respective group. But I'll move the methods up to fit our style.

Reply:
I've got the same "problem" because I, too, like the grouping. So, we keep the styleguide conform stuff for now but discuss changing the rule (inlcuding pro/con) down the road?